packages/astro/src/core/app/index.ts

```diff
@@ -307,6 +264,63 @@ export class App {
 		}
 	}
 
+	/**
+	 * If is a known error code, try sending the according page (e.g. 404.astro / 500.astro).
+	 * This also handles pre-rendered /404 or /500 routes
+	 */
+	async #renderError(request: Request, { routeData, status, response: originalResponse }: RenderErrorOptions) {
+		const errorRouteData = matchRoute('/' + status, this.#manifestData);
+		const url = new URL(request.url);
+		if (errorRouteData) {
+			if (errorRouteData.prerender && !errorRouteData.route.endsWith(`/${status}`)) {
+				const statusURL = new URL(`${this.#baseWithoutTrailingSlash}/${status}`, url);
+				const response = await fetch(statusURL.toString());
+				return this.#mergeResponses(response, originalResponse);
+			}
+			const finalRouteData = routeData ?? errorRouteData;
+			const mod = await this.#getModuleForRoute(errorRouteData);
+			try {
+				const newRenderContext = await this.#createRenderContext(
+					url,
+					request,
+					finalRouteData,
+					mod,
+					status
+				);
+				const page = (await mod.page()) as any;
+				const response = await tryRenderRoute(
+					'page', // this is hardcoded to ensure proper behavior for missing endpoints
+					newRenderContext,
+					this.#env,
+					page
+				) as Response;
+				return this.#mergeResponses(response, originalResponse);
+			} catch {}
+		}
+
+		const response = this.#mergeResponses(new Response(null, { status }), originalResponse);
+		Reflect.set(response, responseSentSymbol, true);
+		return response;
+	}
+
+	#mergeResponses(newResponse: Response, oldResponse?: Response) {
+		if (!oldResponse) return newResponse;
+		const { status, statusText, headers } = oldResponse;
+
+		return new Response(newResponse.body, {
+			status: status === 200 ? newResponse.status : status,
```

Comment:
This is weird; why do we return `newRespnose.status` if we check the original `staus`? Can you add a comment?